File: Exercise_2_MinStack.java
// Time Complexity : push(), pop(), top() , getMin()   -> O(1) for each element
// Space Complexity : O(n) where n is size of the stack and minStack (temporary stack used)
// Did this code successfully run on Leetcode : Yes
/* Any problem you faced while coding this : None
 */
import java.util.Stack;
class MinStack {
    
    Stack<Integer> stack;
    Stack<Integer> minStack;
	// Your code here along with comments explaining your approach
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();  // temporary stack to store the minimum element found till now
    }
    
    public void push(int x) {
        stack.push(x);
        if(minStack.empty() || minStack.peek() >= x){
            minStack.push(x);    // If the new element x is lesser or equal than what it has already, push it
        }
        
    }
    
    public void pop() {
        int out = stack.pop();
        if(out == minStack.peek()){ // If the element getting removed = top of min stack, min stack loses its top element
            minStack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();   // top of minstack represents the min element found till now
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Exercise 2_Min_Stack.py
# Time Complexity : O(1)
# Space Complexity : O(2n) --> O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


class MinStack:

    def __init__(self):
        self.stack = []
        self.min_value = float('inf')

    def push(self, val: int) -> None:
        
        if val <= self.min_value:
            self.stack.append(self.min_value)
            self.min_value = val

        self.stack.append(val)

    def pop(self) -> None:
        
        val = self.stack.pop(-1)
        if self.min_value == val:
            self.min_value = self.stack.pop(-1)

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_value



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()